tenantcostclient: add tenant.sql_usage.estimated_replication_bytes metric

jaylim-crl opened the pull request:
This commit adds a new `tenant.sql_usage.estimated_replication_bytes` metric
to represent the estimated number of bytes for KV replication traffic. Note
that this metric uses the same labeling approach as the network statistics
metrics (i.e. metrics are labeled based on the common prefix and the first
differing part). We will use this, together with the network statistics
metrics, to calculate cross-zone and cross-region network usage.

Epic: https://cockroachlabs.atlassian.net/browse/CC-28725

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126805)
<!-- Reviewable:end -->

andy-kimball commented on pkg/ccl/multitenantccl/tenantcostclient/metrics.go:
```diff
@@ -175,3 +230,68 @@ func (m *metrics) getConsumption(consumption *kvpb.TenantConsumption) {
+	i := 0
+	for i < len(baseLocality.Tiers) {
+		if matching {
+			if i >= len(fromLocality.Tiers) ||
```
What's the benefit of considering `baseLocality`? Why not just consider `fromLocality` and `toLocality`? What scenario are you thinking about here?

andy-kimball commented on pkg/ccl/multitenantccl/tenantcostclient/metrics.go:
```diff
@@ -175,3 +230,68 @@ func (m *metrics) getConsumption(consumption *kvpb.TenantConsumption) {
+			i++
+		} else {
+			// Ensure that the remaining slots are filled with empty strings.
+			fromVals = append(fromVals, "")
```
What benefit does filling remaining slots give?

andy-kimball commented on pkg/kv/kvclient/kvcoord/dist_sender.go:
```diff
@@ -3225,6 +3226,41 @@ func (ds *DistSender) computeNetworkCost(
+// computeWriteReplicationNetworkPaths computes all the network paths taken by
+// the leaseholder when it replicates writes to the other replicas for the given
+// range.
+func (ds *DistSender) computeWriteReplicationNetworkPaths(
```
I'd had in mind that we'd directly increment the write replication metrics here, rather than adding it to a `RequestInfo`. After all, after Andrew's changes, all the other network traffic will be directly recorded in metrics rather than being funneled through the controller. Why should this network metric be special?

jaylim-crl reviewed (commented):
TFTR! I'll convert this to a draft while I rework the approach.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126805#-:-O1IFtD11qA3Pb9f7RhJ:b-ldf8mb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/metrics.go` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126805#gh-1668852138:-O1IGNrw7JzVQodo5N8H:b-y1vk02) ([raw file](https://github.com/cockroachdb/cockroach/blob/48e0a1447ac927b085f174220fed8f9c728a6be1/pkg/ccl/multitenantccl/tenantcostclient/metrics.go#L274)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What's the benefit of considering `baseLocality`? Why not just consider `fromLocality` and `toLocality`? What scenario are you thinking about here?
</blockquote></details>

`baseLocality` is used to determine the set of label keys that the metric is using. We can avoid using `baseLocality`, but that also means that we'll need to initialize the metric lazily, which isn't clean b/c we need to add the metric to the registry as well. 
___
*[`pkg/ccl/multitenantccl/tenantcostclient/metrics.go` line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126805#gh-1668857692:-O1IFuOi512GdxG06Bub:b-qagi5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/48e0a1447ac927b085f174220fed8f9c728a6be1/pkg/ccl/multitenantccl/tenantcostclient/metrics.go#L291)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What benefit does filling remaining slots give?
</blockquote></details>

My initial impression is that we need to ensure the total number of values matches the number of labels the metrics have. I'd see an argument of defaulting these values to empty strings automatically by the metrics library, but I'd have to verify that. Regardless, even if we don't pad the remaining slots (i.e. toVals), we'll still need to pad fromVals since we're combining both fromVals and toVals below. The ordering of label values have to match the ordering of keys.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3232 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126805#gh-1668884450:-O1IGnaO8tTDGji7mfTt:b6e7lzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/48e0a1447ac927b085f174220fed8f9c728a6be1/pkg/kv/kvclient/kvcoord/dist_sender.go#L3232)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd had in mind that we'd directly increment the write replication metrics here, rather than adding it to a `RequestInfo`. After all, after Andrew's changes, all the other network traffic will be directly recorded in metrics rather than being funneled through the controller. Why should this network metric be special?
</blockquote></details>

Discussed offline. We'll increment the metric directly here to avoid needing to create a list of `tenantcostmodel.LocalityNetworkPath` objects.


<!-- Sent from Reviewable.io -->

jaylim-crl committed:
tenantcostclient: add tenant.sql_usage.estimated_replication_bytes metric

This commit adds a new `tenant.sql_usage.estimated_replication_bytes` metric
to represent the estimated number of bytes for KV replication traffic. Note
that this metric uses the same labeling approach as the network statistics
metrics (i.e. metrics are labeled based on the common prefix and the first
differing part). We will use this, together with the network statistics
metrics, to calculate cross-zone and cross-region network usage.

Epic: https://cockroachlabs.atlassian.net/browse/CC-28725

Release note: None

jaylim-crl reviewed (commented):
OK - this is ready for another round of review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126805#-:-O1J8iSM2DBtOoB9SxRb:b-6aif1m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)



<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/kv/kvclient/kvcoord/dist_sender.go:
```diff
@@ -452,7 +461,38 @@ type DistSenderRangeFeedMetrics struct {
+		// cachedPathMetrics stores a cache of network paths to the metrics
+		// which have been initialized. Having this layer of caching prevents us
+		// from needing to compute the normalized locality on every request.
+		cachedPathMetrics map[networkPath]networkPathMetrics
```
This is the rare appropriate time to use `sync.Map`. Currently `updateEstimatedReplicationBytes` is grabbing the mutex once for each replica. But using a sync map would allow avoiding mutex acquisition if the entry is already present.

JeffSwenson commented on pkg/kv/kvclient/kvcoord/dist_sender.go:
```diff
@@ -596,6 +637,101 @@ func makeDistSenderRangeFeedMetrics() DistSenderRangeFeedMetrics {
+func makeLocalityLabelValues(baseLocality, fromLocality, toLocality roachpb.Locality) []string {
+	labelValues := make([]string, 2*len(baseLocality.Tiers))
+	for i := 0; i < len(baseLocality.Tiers); i++ {
+		if i >= len(fromLocality.Tiers) ||
```
nit/consider: You can pull the bounds check into the loop. That makes the break condition easier to read.

```go
localityTiers := min(len(baseLocality.Tiers), len(fromLocality.Tiers), len(toLocality.Tiers))
for i := 0; i < localityTiers; i++ {
    base := baseLocality.Tiers[i].Key
    if fromLocality.Tiers[i].Key != base || toLocality.Tiers[i].Key != base {
		break
	}
    ...
}
```

jaylim-crl reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126805#-:-O1JXXIc3Q17bCO7gtAd:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 473 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126805#gh-1669364924:-O1JXYc3DeCB9tdoT4Mt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab9115cf3b4e4f2da11cce39a8a2b3ac93f13de/pkg/kv/kvclient/kvcoord/dist_sender.go#L473)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

This is the rare appropriate time to use `sync.Map`. Currently `updateEstimatedReplicationBytes` is grabbing the mutex once for each replica. But using a sync map would allow avoiding mutex acquisition if the entry is already present.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 718 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/126805#gh-1669378285:-O1JX_u466yerdGMN2sz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab9115cf3b4e4f2da11cce39a8a2b3ac93f13de/pkg/kv/kvclient/kvcoord/dist_sender.go#L718)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

nit/consider: You can pull the bounds check into the loop. That makes the break condition easier to read.

```go
localityTiers := min(len(baseLocality.Tiers), len(fromLocality.Tiers), len(toLocality.Tiers))
for i := 0; i < localityTiers; i++ {
    base := baseLocality.Tiers[i].Key
    if fromLocality.Tiers[i].Key != base || toLocality.Tiers[i].Key != base {
		break
	}
    ...
}
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/kv/kvclient/kvcoord/dist_sender.go:
```diff
@@ -596,6 +639,113 @@ func makeDistSenderRangeFeedMetrics() DistSenderRangeFeedMetrics {
+// exist in the given base locality. See test cases for more information.
+func makeLocalityLabelValues(baseLocality, fromLocality, toLocality roachpb.Locality) []string {
+	labelValues := make([]string, 2*len(baseLocality.Tiers))
+	length := int(math.Min(
```
As of golang 1.21 there is a `min` built in that you can use here and it doesn't requiring the int/float casts.

https://pkg.go.dev/builtin#min

jaylim-crl replied:
Done.

jeffswenson reviewed (approved):
LGTM

jaylim-crl commented:
TFTR!

jaylim-crl commented:
Discussed offline with @andy-kimball. We're going back to the original approach in this PR (i.e. we're incrementing the metric in OnResponseWait). AndyK will put up a subsequent PR to get rid of `tenantcostmodel.RequestInfo` and `tenantcostmodel.ResponseInfo`.

andy-kimball reviewed (approved):
LGTM

jeffswenson reviewed (approved):
LGTM

jaylim-crl commented:
TFTRs!

jaylim-crl commented:
bors r+
